src/ipnet.rs

```diff
@@ -0,0 +1,828 @@
+        match self {
+            IpNet::V4(inner) => write!(f, "{}", inner),
+            IpNet::V6(inner) => write!(f, "{}", inner),
+        }
+    }
+}
+
+impl std::str::FromStr for IpNet {
+    type Err = IpNetParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let Some(ii) = s.find('/') else {
+            return Err(IpNetParseError::NoPrefix);
+        };
+
+        let addr_str = &s[..ii];
+        let prefix_str = &s[ii + 1..];
+
+        let prefix = prefix_str.parse().map_err(IpNetParseError::InvalidPrefix)?;
+        let addr = addr_str.parse().map_err(IpNetParseError::InvalidAddr)?;
+        IpNet::new(addr, prefix).map_err(IpNetParseError::PrefixValue)
+    }
+}
+
+#[cfg(feature = "schemars")]
+impl schemars::JsonSchema for IpNet {
+    fn schema_name() -> String {
+        "IpNet".to_string()
+    }
+
+    fn json_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+        use crate::schema_util::label_schema;
+        schemars::schema::SchemaObject {
+            subschemas: Some(Box::new(schemars::schema::SubschemaValidation {
+                one_of: Some(vec![
+                    label_schema("v4", gen.subschema_for::<Ipv4Net>()),
+                    label_schema("v6", gen.subschema_for::<Ipv6Net>()),
+                ]),
+                ..Default::default()
+            })),
+            extensions: crate::schema_util::extension("IpNet", "0.1.0"),
+            ..Default::default()
+        }
+        .into()
+    }
+}
+
+/// The maximum width of an IPv4 subnet
+pub const IPV4_NET_WIDTH_MAX: u8 = 32;
+
+/// An IPv4 subnet
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Ipv4Net {
+    addr: Ipv4Addr,
+    width: u8,
+}
+
+impl Ipv4Net {
+    /// Create an Ipv4Net with the given address and prefix width.
+    pub fn new(addr: Ipv4Addr, width: u8) -> Result<Self, IpNetPrefixError> {
+        if width > IPV4_NET_WIDTH_MAX {
+            Err(IpNetPrefixError(width))
+        } else {
+            Ok(Self { addr, width })
+        }
+    }
+
+    /// Create an Ipv4Net that contains *exclusively* the given address.
+    pub fn host_net(addr: Ipv4Addr) -> Self {
+        Self {
+            addr,
+            width: IPV4_NET_WIDTH_MAX,
+        }
+    }
+
+    /// Return the base address used to create this subnet.
+    pub fn addr(&self) -> Ipv4Addr {
+        self.addr
+    }
+
+    /// Return the prefix width.
+    pub fn width(&self) -> u8 {
+        self.width
+    }
+
+    pub(crate) fn mask(&self) -> u32 {
+        u32::MAX << (IPV4_NET_WIDTH_MAX - self.width)
+    }
+
+    /// Return true iff the subnet contains only the base address i.e. the
+    /// size is exactly one address.
+    pub fn is_host_net(&self) -> bool {
+        self.width == IPV4_NET_WIDTH_MAX
+    }
+
+    /// Return the number of addresses contained within this subnet.
+    pub fn size(&self) -> u32 {
+        1u32 << (IPV4_NET_WIDTH_MAX - self.width)
```

Comment:
Both this and the `mask()` function will overflow and panic if the width is 0, since a left-shift with 32 bits is out of range. I think we want to do two things here:

- Write the mask function as:
    ```rust
    pub(crate) fn mask(&self) -> u32 {
        u32::MAX.checked_shl(IPV4_NET_WIDTH_MAX - self.width).unwrap_or(u32::MAX)
    }
    ```
    That could be written differently, like an explicit check on `self.width`, if you prefer. I'm not wedded to that specific implementation.
- Write `size()` using the bitwise complement of `mask()`

See [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a3cfcb315941861019af85cd42b52af8) for an example, including where it panics.

Reply:
Great catch! Should that be
```rust
    u32::MAX.checked_shl(IPV4_NET_WIDTH_MAX - self.width).unwrap_or(0)
```

I'm not 100% sure what `0.0.0.0/0` means, but I imagine it means EVERYTHING i.e. mask is `0` (i.e. whereas if the width is 32, the mask is `u32::MAX`)

Reply:
Yeah, that's correct. `0.0.0.0/0` is the prefix that contains everything, sometimes referred to as the default route.